pam/pam.go

```diff
@@ -69,7 +82,25 @@ func sendReturnMessageToPam(mTx pam.ModuleTransaction, retStatus adapter.PamRetu
 
 // Authenticate is the method that is invoked during pam_authenticate request.
 func (h *pamModule) Authenticate(mTx pam.ModuleTransaction, flags pam.Flags, args []string) error {
-	return h.handleAuthRequest(authd.SessionMode_AUTH, mTx, flags, args)
+	// Do not try to start authentication again if we've been already through this.
+	// Since PAM modules can be stacked, so we may suffer reentry that is fine but it should
+	// be explicitly allowed.
+	_, err := mTx.GetData(authenticatedKey)
+	if err == nil && parseArgs(args)["force_reauth"] != "true" {
```

Comment:
I would do stricter arg parsing in `parseArgs`, at least, warning on unknown keys. That would at least document in one place our supported keys.

Reply:
Ok.... This is done now, but for testing it I had to steal few commits from #233

Reply:
You didn't need to assert on the logging messages. IMHO, the test for it should only be in the sense of "Unsupported argument gets ignored" and then proceed with a simple authentication (rather than stopping in the middle of nowhere as the current tape file does).

Reply:
Ok, done... I feel it's incomplete testing, but I also agree checking on logs isn't the best.

At least, it would have been better with a proper log handler and ensure that we call it, but that's fine for now.